runtime/nvqir/NVQIR.cpp

```diff
@@ -629,49 +629,37 @@ void releasePackedQubitArray(Array *a) {
   return;
 }
 
-/// @brief Utility function used by Quake->QIR to invoke a QIR QIS function
-/// with a variadic list of control qubits.
-/// @param nControls
-/// @param QISFunction
-/// @param
-void invokeWithControlQubits(const std::size_t nControls,
-                             void (*QISFunction)(Array *, Qubit *), ...) {
-  // Create the Control Array *, This should
-  // be deallocated upon function exit.
-  auto ctrlArray = std::make_unique<Array>(nControls, sizeof(std::size_t));
-
-  // Start up the variadic arg processing
-  va_list args;
-  va_start(args, QISFunction);
-
-  std::size_t nSetPointers = 0;
-  while (nSetPointers < nControls) {
-    // For each variadic arg, get it (its a Qubit*)
-    // and set it on the array.
-    Qubit *ctrli = va_arg(args, Qubit *);
-    auto ctrliRawPtr =
-        __quantum__rt__array_get_element_ptr_1d(ctrlArray.get(), nSetPointers);
-    cudaq::info("before reinterpret cast");
-    *reinterpret_cast<Qubit **>(ctrliRawPtr) = ctrli;
-    nSetPointers++;
+/// This is the generalized version of invoke that does not use a va_list
+/// argument. It provides a general interface to allow invoking a general
+/// quantum operation, which may contain some number of rotation arguments
+/// (double), control arguments (either qubits or arrays), and target arguments
+/// (qubits). \p numRotationOperands and \p numTargetOperands must be no more
+/// than 2. \p numTargetOperands must be at least 1. The arguments are passed as
+/// arrays (built by the caller on the stack) as: \p params, \p controls, and \p
+/// targets. \p isArrayAndLength is a buffer used to determine the type of the
+/// control arguments and must be present if \p numControlOperands is non-zero.
+/// The length of \p isArrayAndLength must also be \p numControlOperands.
+static void commonInvokeWithRotationsControlsTargets(
+    std::size_t numRotationOperands, double *params,
+    std::size_t numControlOperands, std::size_t *isArrayAndLength,
+    Qubit **controls, std::size_t numTargetOperands, Qubit **targets,
+    void (*QISFunction)()) {
+  if (numRotationOperands > 2) {
+    cudaq::info(
+        "[runtime] FATAL ERROR. Invoke has invalid number of rotations, {}.",
+        numRotationOperands);
+    return;
   }
+  if (numTargetOperands < 1 || numTargetOperands > 2) {
+    cudaq::info(
+        "[runtime] FATAL ERROR. Invoke has invalid number of targets, {}.",
+        numTargetOperands);
```

Comment:
A user won't see this fatal error unless they are running with the logging environment variable, so it may be better to throw a runtime error that they'd be guaranteed to see if it really is fatal.